### PR TITLE
Rachel DiCesare - Unicorn

### PR DESCRIPTION
**Pull request by RachelDiCesare93**
### Day 1: Get `print8.ls8` running
- [x] Inventory what is here
- [x] Implement `struct cpu` in `cpu.h`
- [x] Add RAM functions
- [x] Implement `cpu_init()`
- [x] Implement the core of `cpu_run()`
- [x] Implement the `HLT` instruction handler
- [x] Add the `LDI` instruction
- [x] Add the `PRN` instruction

### Day 2: Get `mult.ls8` running, load files dynamically
- [x] Un-hardcode the machine code
- [x] Implement a Multiply and Print the Result (run `mult8.ls8`)
- [ ] Beautify your `cpu_run()` function, if needed

### Day 3: Get `stack.ls8` running
- [x] Implement System Stack

### Day 4: Get `call.ls8` running
- [x] Implement Subroutine Calls

### Stretch
- [ ] Add the timer interrupt to the LS-8 emulator
- [ ] Add the keyboard interrupt to the LS-8 emulator

### CODE REVIEW

File: ls8/Step0.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+- asm/mult.asm: multiplies two numbers
+- asm/print8.asm: loads a number into first register and prints it out
+- asm/printstr.asm: prints a string at a given address
+- asm/README.md: details what the assembler does
+- asm/sctest.asm: tests the Sprint Challenge
+- asm/stack.asm: tests the stack functionality
+
+
+
+
```
Comment (Mister-Corn) [inline ls8/Step0.md:34]: :heart_eyes:

File: ls8/cpu.c [hunk 2]
```diff
@@ -50,9 +62,36 @@ void cpu_run(struct cpu *cpu)
...
+        printf("%d\n", cpu->registers[(int)operandA]);
+        cpu->PC += 2;
+        break;
+      default:
+        printf("%d", instruction);
+        printf("Default case reached.\n");
+        break;
+    }
     // 4. Move the PC to the next instruction.
+    // cpu->PC += (int)(instruction >> 6);
```
Comment (Mister-Corn) [inline ls8/cpu.c]: In actuality, the code above is sound. There's one thing: if you notice the workaround code above, notice that we move the `PC` 2 for `PRN`, even though there 's only one operand after PRN, and 3 for `LDI`, despite being two operands. The reason we're moving it one more is that the PC needs to be pointing to the next instruction to be executed after the PC moves, but if you just move it by the number of operands it has, you'll find yourself in the wrong place, often the operand before the next instruction.

TL;DR 
```suggestion
    cpu->PC += (int)(instruction >> 6) + 1;
```

And remove the hard-coded PC incrementors above.

File: ls8/cpu.c [hunk 3]
```diff
@@ -50,9 +62,36 @@ void cpu_run(struct cpu *cpu)
   while (running) {
     // TODO
     // 1. Get the value of the current instruction (in address PC).
+    int PC = cpu->PC;
+    unsigned char instruction = cpu_ram_read(cpu, PC);
```
Comment (Mister-Corn) [inline ls8/cpu.c]: The way we did this yesterday is totally fine. If you want to be faithful to the hardware we are emulating, you can add an `IR` member in the cpu struct. In the LS8 spec, `IR` is the instruction register: the register which holds a copy of the current instruction being executed. In a similar vein, the `IR` member in the struct can serve as the place to store the instruction before "execution".

**Review by Mister-Corn — COMMENTED** [review]
Great job, Rachel. :heart_eyes: You're doing well. I hope the repo is helping you explore how CPU's do their thang.

File: ls8/cpu.c [hunk 4]
```diff
@@ -73,51 +74,89 @@ void cpu_run(struct cpu *cpu)
...
     // 2. switch() over it to decide on a course of action.
     // 3. Do whatever the instruction should do according to the spec.
     switch(IR) {
       case HLT:
-        printf("HLT\n");
+        #if DEBUG
+          printf("HLT\n");
+          #endif
         running = 0;
+        INC_PC = 1;
```
Comment (Mister-Corn) [inline ls8/cpu.c]: Similar to how we are dynamically change the PC, there's a bit in the OPCODE that signals whether the operation directly manipulates the program counter (like `RET` and `CALL`). You could use that to condition your PC change. Check the spec for more details!

File: ls8/cpu.c [hunk 5]
```diff
@@ -33,13 +38,28 @@ void alu(struct cpu *cpu, enum alu_op op, unsigned char regA, unsigned char regB
 {
   switch (op) {
     case ALU_MUL:
-      // TODO
+      cpu->registers[regA] = cpu->registers[regA] * cpu->registers[regB];
```
Comment (Mister-Corn) [inline ls8/cpu.c]: This is something that I didn't notice before, but Kyle mentioned this lecture. There might be problems if the result of the operation is greater than 255 (0xFF), so you can contain the answer to be within that range by bit masking the answer.

Try:
```suggestion
      cpu->registers[regA] = (cpu->registers[regA] * cpu->registers[regB]) & 0xFF;
```

See if that changes the operation of your emulator. This applies to all operations in the ALU.

**Review by Mister-Corn — COMMENTED** [review]
Good work, Rachel. As I mentioned before, it might be a good time to reflect and inventory the work you've done and see how everything is jivin' together. Again, well done. 👍
